File: src/lib/srs/sm2.ts
// src/lib/srs/sm2.ts
// SM-2 (SuperMemo 2) spaced repetition algorithm.
//
// Pure functions with no I/O so the exact same logic can run in the browser
// today and be ported to Swift for the iOS app later. Dates cross this
// boundary as ISO strings (what Supabase returns) or Date objects.

// Quality of response (how well the user remembered the card)
export enum ResponseQuality {
  BLACKOUT = 0, // Complete blackout - no memory
  INCORRECT = 1, // Incorrect response with familiar feeling
  HARD = 2, // Correct response with serious difficulty
  GOOD = 3, // Correct response after some hesitation
  EASY = 4, // Perfect response with ease
}

export interface SpacedRepetitionData {
  easeFactor: number // How "easy" this card is (2.5 is default)
  interval: number // Days until next review
  repetitions: number // How many times studied successfully
  nextReviewDate: Date // When to show this card next
}

/**
 * Calculate the next review data based on user response.
 * This is the core of the spaced repetition algorithm.
 */
export function calculateNextReview(
  currentData: SpacedRepetitionData,
  responseQuality: ResponseQuality
): SpacedRepetitionData {
  let { easeFactor, interval, repetitions } = currentData

  // Update ease factor based on response quality:
  // better responses make the card "easier" (longer intervals)
  easeFactor = Math.max(
    1.3,
    easeFactor +
      (0.1 - (5 - responseQuality) * (0.08 + (5 - responseQuality) * 0.02))
  )

  // If response was poor (quality < 3), reset repetitions and set short interval
  if (responseQuality < 3) {
    repetitions = 0
    interval = 1 // Show again tomorrow
  } else {
    repetitions += 1

    if (repetitions === 1) {
      interval = 1 // First successful review: 1 day
    } else if (repetitions === 2) {
      interval = 6 // Second successful review: 6 days
    } else {
      // Subsequent reviews: multiply previous interval by ease factor
      interval = Math.round(interval * easeFactor)
    }
  }

  const nextReviewDate = new Date()
  nextReviewDate.setDate(nextReviewDate.getDate() + interval)

  return { easeFactor, interval, repetitions, nextReviewDate }
}

/** Determine if a card is due for review. */
export function isCardDue(nextReviewDate: string | Date): boolean {
  return new Date(nextReviewDate) <= new Date()
}

/** Create initial spaced repetition data for a new card. */
export function createInitialReviewData(): SpacedRepetitionData {
  return {
    easeFactor: 2.5,
    interval: 0, // New card - show immediately
    repetitions: 0,
    nextReviewDate: new Date(), // Due now
  }
}

/** Get a human-readable description of the next review interval. */
export function getIntervalDescription(interval: number): string {
  if (interval === 0) return "New card"
  if (interval === 1) return "1 day"
  if (interval < 7) return `${interval} days`
  if (interval < 30) return `${Math.round(interval / 7)} week(s)`
  if (interval < 365) return `${Math.round(interval / 30)} month(s)`
  return `${Math.round(interval / 365)} year(s)`
}

File: ios/CantoneseLearner/Sources/Core/SM2.swift
import Foundation

/// Line-for-line port of `src/lib/srs/sm2.ts`.
enum ResponseQuality: Int, CaseIterable, Sendable {
    case blackout = 0, incorrect = 1, hard = 2, good = 3, easy = 4
}

struct SpacedRepetitionData: Sendable {
    var easeFactor: Double
    var interval: Int       // days
    var repetitions: Int
    var nextReviewDate: Date
}

enum SM2 {
    static func createInitialReviewData(now: Date = Date()) -> SpacedRepetitionData {
        SpacedRepetitionData(easeFactor: 2.5, interval: 0, repetitions: 0, nextReviewDate: now)
    }

    static func calculateNextReview(_ current: SpacedRepetitionData, quality q: ResponseQuality, now: Date = Date()) -> SpacedRepetitionData {
        var easeFactor = current.easeFactor
        var interval = current.interval
        var repetitions = current.repetitions

        // q is 0..4 (NOT 0..5): (5 - q) is used verbatim.
        let d = Double(5 - q.rawValue)
        easeFactor = max(1.3, easeFactor + (0.1 - d * (0.08 + d * 0.02)))

        if q.rawValue < 3 {
            repetitions = 0
            interval = 1
        } else {
            repetitions += 1
            if repetitions == 1 {
                interval = 1
            } else if repetitions == 2 {
                interval = 6
            } else {
                // JS Math.round: half toward +inf. Positive values → half away from zero.
                interval = Int((Double(interval) * easeFactor).rounded(.toNearestOrAwayFromZero))
            }
        }

        // JS: setDate(getDate() + interval) — local calendar-day arithmetic, keeps time of day.
        let next = Calendar.current.date(byAdding: .day, value: interval, to: now) ?? now.addingTimeInterval(Double(interval) * 86400)
        return SpacedRepetitionData(easeFactor: easeFactor, interval: interval, repetitions: repetitions, nextReviewDate: next)
    }

    static func isCardDue(_ nextReviewDate: Date, now: Date = Date()) -> Bool {
        nextReviewDate <= now
    }

    static func intervalDescription(_ interval: Int) -> String {
        if interval == 0 { return "New card" }
        if interval == 1 { return "1 day" }
        if interval < 7 { return "\(interval) days" }
        if interval < 30 { let w = Int((Double(interval) / 7).rounded()); return "\(w) week\(w == 1 ? "" : "s")" }
        if interval < 365 { let m = Int((Double(interval) / 30).rounded()); return "\(m) month\(m == 1 ? "" : "s")" }
        let y = Int((Double(interval) / 365).rounded()); return "\(y) year\(y == 1 ? "" : "s")"
    }
}
